File: spec/factories/permissions.rb
FactoryGirl.define do

  factory :collection_api_client_permission, class: Permissions::CollectionApiClientPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/2.0}
    edit_metadata_and_relations {FactoryHelper.rand_bool 1/4.0}

    api_client {ApiClient.find_random || (FactoryGirl.create :api_client)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    collection {Collection.find_random || (FactoryGirl.create :collection)}

  end



  factory :collection_group_permission, class: Permissions::CollectionGroupPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/2.0}
    edit_metadata_and_relations {FactoryHelper.rand_bool 1/4.0}

    group {Group.find_random || (FactoryGirl.create :group)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    collection {Collection.find_random || (FactoryGirl.create :collection)}

  end

  factory :filter_set_group_permission, class: Permissions::FilterSetGroupPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/2.0}
    edit_metadata_and_filter {FactoryHelper.rand_bool 1/4.0}

    group {Group.find_random || (FactoryGirl.create :group)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    filter_set {Collection.find_random || (FactoryGirl.create :filter_set)}

  end


  factory :filter_set_api_client_permission, class: Permissions::FilterSetApiClientPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/2.0}
    edit_metadata_and_filter {FactoryHelper.rand_bool 1/4.0}

    api_client {ApiClient.find_random || (FactoryGirl.create :api_client)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    filter_set {FilterSet.find_random || (FactoryGirl.create :filter_set)}

  end


  factory :filter_set_user_permission, class: Permissions::FilterSetUserPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool}
    edit_metadata_and_filter {FactoryHelper.rand_bool 1/2.0}
    edit_permissions {FactoryHelper.rand_bool}

    user {User.find_random || (FactoryGirl.create :user)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    filter_set {FilterSet.find_random || (FactoryGirl.create :filter_set)}

  end



  factory :media_entry_user_permission, class: Permissions::MediaEntryUserPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/4.0}
    get_full_size{get_metadata_and_previews and FactoryHelper.rand_bool}
    edit_metadata {FactoryHelper.rand_bool 1/4.0}
    edit_permissions {edit_metadata and FactoryHelper.rand_bool}

    user {User.find_random || (FactoryGirl.create :user)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    media_entry {MediaEntry.find_random || (FactoryGirl.create :media_entry)}

  end


  factory :collection_user_permission, class: Permissions::CollectionUserPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/4.0}
    edit_metadata_and_relations {FactoryHelper.rand_bool 1/4.0}

    user {User.find_random || (FactoryGirl.create :user)}
    updator {User.find_random || (FactoryGirl.create :user)}
    collection {Collection.find_random || (FactoryGirl.create :collection)}

  end


  factory :media_entry_group_permission, class: Permissions::MediaEntryGroupPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/4.0}
    get_full_size{get_metadata_and_previews and FactoryHelper.rand_bool}
    edit_metadata {FactoryHelper.rand_bool 1/4.0}

    group {Group.find_random || (FactoryGirl.create :group)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    media_entry {MediaEntry.find_random || (FactoryGirl.create :media_entry)}

  end



  factory :media_entry_api_client_permission, class: Permissions::MediaEntryApiClientPermission do

    get_metadata_and_previews {FactoryHelper.rand_bool 1/4.0}
    get_full_size{get_metadata_and_previews and FactoryHelper.rand_bool}

    api_client {ApiClient.find_random || (FactoryGirl.create :api_client)} 
    updator {User.find_random || (FactoryGirl.create :user)}
    media_entry {MediaEntry.find_random || (FactoryGirl.create :media_entry)}

  end



end

File: app/javascript/models/media-entry/permissions.js
/*
 * decaffeinate suggestions:
 * DS102: Remove unnecessary code created because of implicit returns
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/main/docs/suggestions.md
 */
import Collection from 'ampersand-rest-collection'
import AppResource from '../shared/app-resource.js'
import ResourcePermissions from '../shared/resource-permissions.js'
import User from '../user.js'
import Group from '../group.js'
import ApiClient from '../api-client.js'

// NOTE: 'trilean' type for usage in batch - can be true, false or mixed

// Child Collections/Models (defined here because they are not needed anywhere else)
const MediaEntryPublicPermission = AppResource.extend({
  type: 'MediaEntryPublicPermission',
  props: {
    get_metadata_and_previews: ['trilean'],
    get_full_size: ['trilean'],
    tooltip_text: { type: 'string', default: '', allowNull: true }
  }
})

const MediaEntryUserPermissions = Collection.extend({
  model: AppResource.extend({
    type: 'MediaEntryUserPermission',
    children: {
      subject: User
    },
    props: {
      get_metadata_and_previews: ['trilean', false, false],
      get_full_size: ['trilean', false, false],
      edit_metadata: ['trilean', false, false],
      edit_permissions: ['trilean', false, false]
    }
  })
})

const MediaEntryGroupPermissions = Collection.extend({
  type: 'MediaEntryGroupPermissions',
  model: AppResource.extend({
    type: 'MediaEntryGroupPermission',
    children: {
      subject: Group
    },
    props: {
      get_metadata_and_previews: ['trilean', false, false],
      get_full_size: ['trilean', false, false],
      edit_metadata: ['trilean', false, false]
    }
  })
})

const MediaEntryApiClientPermissions = Collection.extend({
  type: 'MediaEntryApiClientPermissions',
  model: AppResource.extend({
    type: 'MediaEntryApiClientPermission',
    children: {
      subject: ApiClient
    },
    props: {
      get_metadata_and_previews: ['trilean', false, false],
      get_full_size: ['trilean', false, false]
    }
  })
})

module.exports = ResourcePermissions.extend({
  type: 'MediaEntryPermissions',

  children: {
    // public permission is just 1 subject, so not a collection!
    public_permission: MediaEntryPublicPermission
  },

  collections: {
    user_permissions: MediaEntryUserPermissions,
    group_permissions: MediaEntryGroupPermissions,
    api_client_permissions: MediaEntryApiClientPermissions
  },

  // custom serialize to match what rails expects — used on this.save()
  serialize(data) {
    return { media_entry: AppResource.prototype.serialize.call(this, data) }
  }
})
